autorel/utils/docker_cli.py

```diff
@@ -0,0 +1,150 @@
+from docker import Client
+
+
+class DockerException(Exception):
+    """
+        Exceptions corresponding to Docker
+        class
+    """
+    def __init__(self, message):
+        super(self, DockerException).__init__(self)
+        self._message = message
+
+    def __str__(self):
+        return self._message
+
+
+class Docker(object):
+    """
+        Build the source code in a docker environment using the
+        supplied builder instance
+    """
+    def __init__(self):
+        self._cli = Client(base_url="unix://var/run/docker.sock",
+                           version="auto"
+                           )
+        # container running flag
+        self._running = False
+        # configure logger
+        self._logger = logging.getLogger(__name__)
+        self._logger.setLevel(logging.DEBUG)
+        channel = logging.StreamHandler(sys.stdout)
+        channel.setLevel(logging.DEBUG)
+        formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
+        channel.setFormatter(formatter)
+        self._logger.addHandler(channel)
+
+    def _cli_command(self, command_, *args, response=False, **kwargs):
+        """
+            Runs a docker client command and returns output if response
+            flag is set to be True
+        """
+        self._logger.debug("Running {0} command on docker client"
+            .format(command_))
+        command = getattr(self._cli, command_)
+        if response:
+            result = command(*args, **kwargs)
+            if isinstance(result, bytes):
+                result = result.decode('utf-8')
+            self._logger.debug(result)
+            return result
+        else:
+            # set the output of client command to be a generator
+            kwargs['stream'] = True
+            for line in command(*args, **kwargs):
+                line = line.decode('utf-8')
+                self._logger.debug(line)
+
+    def _exec_in_container(self, command_):
+        """
+            Executes a command in a running docker container
+            and checks for the execution status
+        """
+        cmd = 'bash -c \"{0}\"'.format(command_)
+        self._logger.debug("Executing in container ({0}) : {1}"
+            .format(self._container_id,cmd))
+        exec_instance = self._cli_command("exec_create",
+                                          response=True,
+                                          container=self._container_id,
+                                          cmd=cmd
+                                          )
+        exec_id = exec_instance.get('Id')
+        self._cli_command("exec_start",
+                          exec_id
+                          )
+        exec_status = self._cli_command("exec_inspect",
+                                        exec_id,
+                                        response=True
+                                        )
+        if exec_status["ExitCode"] != 0:
+            raise CommandFailure(response)
```

Comment:
Need to fix this. I did some refactoring after a test run. You please do an architectural analysis at this time. I will test run and fix these after incorporating your comments.